@vitejs/plugin-legacy rollupOptions.output.entryFileNames and rollupOptions.output.chunkFileNames does not support the function format.

<!-- 中文用户请注意：请仔细阅读以下模版，如果不遵循模版，issue 将会被直接关闭。 -->
<!--
  !!! IMPORTANT !!!
  Please do not ignore this template. If you do, your issue will be closed immediately.
-->

### Describe the bug
i meet this error:
`Error: @vitejs/plugin-legacy rollupOptions.output.entryFileNames and rollupOptions.output.chunkFileNames does not support the function format.`
when i update my dependence '@vitejs/plugin-legacy' from '1.4.0' to '1.4.1'


<!-- A clear and concise description of what the bug is. -->

<!-- If you intend to submit a PR for this issue, tell us in the description. Thanks! -->

### Reproduction
my config: 
```
plugins: [
    legacy()
   ...
],
build: {
    rollupOptions: {
      output: {
         chunkFileNames: ChunkInfo => {
           //do something with ChunkInfo
          return 'assets/'+myName+'.[hash].js'
      }
    }
  }
}
```

<!--
  Please provide a link to a repo that can reproduce the problem you ran into.

  A reproduction is required unless you are absolutely sure that the issue is obvious and the provided information is enough to understand the problem. If a report is vague (e.g. just a generic error message) and has no reproduction, it will receive a "need reproduction" label. If no reproduction is provided after 3 days, it will be auto-closed.
-->

### System Info

Output of `npx envinfo --system --npmPackages vite,@vitejs/plugin-vue --binaries --browsers`:

```node
System:
    OS: macOS 10.15.7
    CPU: (8) x64 Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
    Memory: 1.71 GB / 16.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 14.16.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.6 - /usr/local/bin/npm
  Browsers:
    Chrome: 90.0.4430.212
    Safari: 14.0.3
  npmPackages:
    vite: ^2.3.3 => 2.3.5 
```

Used package manager: <!-- npm | yarn | pnpm -->
`@vitejs/plugin-legacy: 1.4.1`
### Logs <!-- (Optional if provided reproduction) -->

<!--
  Please try not to insert an image but copy paste the log text.

  1. Run `vite` or `vite build` with the `--debug` flag.
  2. Provide the error log here.
     `node` is used as highlight to improve some colors in stack-traces.
     If it doesn't work quite well, try `console`.
-->

```node
error during build:
Error: @vitejs/plugin-legacy rollupOptions.output.entryFileNames and rollupOptions.output.chunkFileNames does not support the function format.
    at getLegacyOutputFileName (/Users/sy/work/syao/activity-front/node_modules/@vitejs/plugin-legacy/index.js:182:17)
    at createLegacyOutput (/Users/sy/work/syao/activity-front/node_modules/@vitejs/plugin-legacy/index.js:206:27)
    at Object.configResolved (/Users/sy/work/syao/activity-front/node_modules/@vitejs/plugin-legacy/index.js:215:33)
    at /Users/sy/work/syao/activity-front/node_modules/vite/dist/node/chunks/dep-01eec7aa.js:70545:127
    at Array.map (<anonymous>)
    at resolveConfig (/Users/sy/work/syao/activity-front/node_modules/vite/dist/node/chunks/dep-01eec7aa.js:70545:35)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async doBuild (/Users/sy/work/syao/activity-front/node_modules/vite/dist/node/chunks/dep-01eec7aa.js:45228:20)
    at async build (/Users/sy/work/syao/activity-front/node_modules/vite/dist/node/chunks/dep-01eec7aa.js:45216:16)
    at async CAC.<anonymous> (/Users/sy/work/syao/activity-front/node_modules/vite/dist/node/cli.js:13928:9)
npm ERR! code ELIFECYCLE
```

---

### Before submitting the issue, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md).
- [x] Read the [docs](https://vitejs.dev/guide).
- [x] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [x] Provide a description in this issue that describes the bug.
- [x] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).

--- a/packages/playground/legacy/__tests__/legacy.spec.ts
+++ b/packages/playground/legacy/__tests__/legacy.spec.ts
@@ -1,4 +1,4 @@
-import { isBuild } from '../../testUtils'
+import { isBuild, untilUpdated } from '../../testUtils'
 
 test('should work', async () => {
   expect(await page.textContent('#app')).toMatch('Hello')
@@ -16,5 +16,31 @@ test('transpiles down iterators correctly', async () => {
 test('wraps with iife', async () => {
   expect(await page.textContent('#babel-helpers')).toMatch(
     'exposed babel helpers: false'
+  )
+})
+
+test('generates assets', async () => {
+  await untilUpdated(
+    () => page.textContent('#assets'),
+    isBuild
+      ? [
+          'index: 404',
+          'index-legacy: 404',
+          'chunk-async: 404',
+          'chunk-async-legacy: 404',
+          'immutable-chunk: 200',
+          'immutable-chunk-legacy: 200',
+          'polyfills-legacy: 404'
+        ].join('\n')
+      : [
+          'index: 404',
+          'index-legacy: 404',
+          'chunk-async: 404',
+          'chunk-async-legacy: 404',
+          'immutable-chunk: 404',
+          'immutable-chunk-legacy: 404',
+          'polyfills-legacy: 404'
+        ].join('\n'),
+    true
   )
 })

--- a/packages/playground/legacy/immutable-chunk.js
+++ b/packages/playground/legacy/immutable-chunk.js
@@ -1,0 +1,18 @@
+const chunks = [
+  'index',
+  'index-legacy',
+  'chunk-async',
+  'chunk-async-legacy',
+  'immutable-chunk',
+  'immutable-chunk-legacy',
+  'polyfills-legacy'
+]
+
+export function fn() {
+  return Promise.all(
+    chunks.map(async (name) => {
+      const response = await fetch(`/assets/${name}.js`)
+      return `${name}: ${response.status}`
+    })
+  )
+}

--- a/packages/playground/legacy/index.html
+++ b/packages/playground/legacy/index.html
@@ -2,4 +2,5 @@
 <div id="env"></div>
 <div id="iterators"></div>
 <div id="babel-helpers"></div>
+<div id="assets"></div>
 <script type="module" src="./main.js"></script>

--- a/packages/playground/legacy/main.js
+++ b/packages/playground/legacy/main.js
@@ -27,6 +27,13 @@ text(
   String.raw`exposed babel helpers: ${window._templateObject != null}`
 )
 
+// dynamic chunk names
+import('./immutable-chunk.js')
+  .then(({ fn }) => fn())
+  .then((assets) => {
+    text('#assets', assets.join('\n'))
+  })
+
 function text(el, text) {
   document.querySelector(el).textContent = text
 }

--- a/packages/playground/legacy/vite.config.js
+++ b/packages/playground/legacy/vite.config.js
@@ -11,7 +11,19 @@ module.exports = {
 
   build: {
     // make tests faster
-    minify: false
+    minify: false,
+
+    rollupOptions: {
+      output: {
+        chunkFileNames(chunkInfo) {
+          if (chunkInfo.name === 'immutable-chunk') {
+            return `assets/${chunkInfo.name}.js`
+          }
+
+          return `assets/chunk-[name].[hash].js`
+        }
+      }
+    }
   },
 
   // special test only hook

--- a/packages/plugin-legacy/index.js
+++ b/packages/plugin-legacy/index.js
@@ -182,7 +182,7 @@ function viteLegacyPlugin(options = {}) {
 
         return (chunkInfo) => {
           let fileName =
-            typeof fileNames == 'function' ? fileNames(chunkInfo) : fileNames
+            typeof fileNames === 'function' ? fileNames(chunkInfo) : fileNames
 
           if (fileName.includes('[name]')) {
             // [name]-[hash].[format] -> [name]-legacy-[hash].[format]

--- a/packages/plugin-legacy/index.js
+++ b/packages/plugin-legacy/index.js
@@ -99,7 +99,7 @@ function viteLegacyPlugin(options = {}) {
     },
 
     async generateBundle(opts, bundle) {
-      if (!isLegacyOutput(opts)) {
+      if (!isLegacyBundle(bundle, opts)) {
         if (!modernPolyfills.size) {
           return
         }
@@ -168,8 +168,9 @@ function viteLegacyPlugin(options = {}) {
       }
 
       /**
-       * @param {string|((chunkInfo: import('rollup').PreRenderedChunk)=>string)} fileNames
+       * @param {string | ((chunkInfo: import('rollup').PreRenderedChunk) => string)} fileNames
        * @param {string?} defaultFileName
+       * @returns {(chunkInfo: import('rollup').PreRenderedChunk) => string)}
        */
       const getLegacyOutputFileName = (
         fileNames,
@@ -179,21 +180,20 @@ function viteLegacyPlugin(options = {}) {
           return path.posix.join(config.build.assetsDir, defaultFileName)
         }
 
-        // does not support custom functions.
-        if (typeof fileNames === 'function') {
-          throw new Error(
-            `@vitejs/plugin-legacy rollupOptions.output.entryFileNames and rollupOptions.output.chunkFileNames` +
-              ` does not support the function format.`
-          )
-        }
+        return (chunkInfo) => {
+          let fileName =
+            typeof fileNames == 'function' ? fileNames(chunkInfo) : fileNames
 
-        let fileName = defaultFileName
-        // Custom string file return format.
-        if (fileNames && typeof fileNames === 'string') {
-          fileName = fileNames.replace(/\[name\]/, '[name]-legacy')
-        }
+          if (fileName.includes('[name]')) {
+            // [name]-[hash].[format] -> [name]-legacy-[hash].[format]
+            fileName = fileName.replace('[name]', '[name]-legacy')
+          } else {
+            // entry.js -> entry-legacy.js
+            fileName = fileName.replace(/(.+)\.(.+)/, '$1-legacy.$2')
+          }
 
-        return fileName
+          return fileName
+        }
       }
 
       /**
@@ -219,7 +219,7 @@ function viteLegacyPlugin(options = {}) {
     },
 
     renderChunk(raw, chunk, opts) {
-      if (!isLegacyOutput(opts)) {
+      if (!isLegacyChunk(chunk, opts)) {
         if (
           options.modernPolyfills &&
           !Array.isArray(options.modernPolyfills)
@@ -399,7 +399,7 @@ function viteLegacyPlugin(options = {}) {
     },
 
     generateBundle(opts, bundle) {
-      if (isLegacyOutput(opts)) {
+      if (isLegacyBundle(bundle, opts)) {
         // avoid emitting duplicate assets
         for (const name in bundle) {
           if (bundle[name].type === 'asset') {
@@ -564,14 +564,27 @@ function polyfillsPlugin(imports) {
 }
 
 /**
+ * @param {import('rollup').RenderedChunk} chunk
  * @param {import('rollup').NormalizedOutputOptions} options
  */
-function isLegacyOutput(options) {
-  return (
-    options.format === 'system' &&
-    typeof options.entryFileNames === 'string' &&
-    options.entryFileNames.includes('-legacy')
-  )
+function isLegacyChunk(chunk, options) {
+  return options.format === 'system' && chunk.fileName.includes('-legacy')
+}
+
+/**
+ * @param {import('rollup').OutputBundle} bundle
+ * @param {import('rollup').NormalizedOutputOptions} options
+ */
+function isLegacyBundle(bundle, options) {
+  if (options.format === 'system') {
+    const entryChunk = Object.values(bundle).find(
+      (output) => output.type === 'chunk' && output.isEntry
+    )
+
+    return !!entryChunk && entryChunk.fileName.includes('-legacy')
+  }
+
+  return false
 }
 
 /**

